### PR TITLE
GH-653b: As a user I need better error reporting for external libraries and first changes to support user access on node_modules folder 

### PR DESCRIPTION
**Pull request by mmews-n4**
follow-up of: #673

**changes**
* changed library manager to detect out-of-sync folder and index
  * synchronization after install/uninstall npms
  * synchronization detects change set
  * update of versions supported
  * "reload libraries" now reloads all external project from scratch and wipes them off the Xtext index first
  * factored out NPM commands
* fixed all package.json files regarding `author` and `repository` information
* made library manger actions cancelable
* added/improved dialogs that show more information about occurred errors
* adjusted dependency management between shared and language (N4JS)
  * introduced delegation to N4JS-language injector
* added fast fail when Guice fails on startup: N4JSGracefulActivator
* library manager supports no-network when uninstalling or installing from cache
* org.eclipse.emf.common.util.URI does not implement `equals()` --> use `URIUtils.equals()`
* Refresh (F5) in the project explorer will check if node_modules is out of sync

### CODE REVIEW

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/ExternalLibrariesActionsHelper.java [hunk 1]
```diff
@@ -38,7 +37,7 @@
 	@Inject
 	private StatusHelper statusHelper;
 	@Inject
-	private NpmManager npmManager;
+	private LibraryManager npmManager;
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/ExternalLibrariesActionsHelper.java]: at some point sync names.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/internal/N4JSExternalLibraryResourceDescriptionsPersisterContribution.java [hunk 2]
```diff
@@ -63,9 +57,10 @@ protected IStatus run(final IProgressMonitor monitor) {
 						@Override
 						public void run(final IProgressMonitor progress) throws CoreException {
 							try {
-								externalLibraryWorkspace.updateState();
-								externalLibrariesReloadHelper.reloadLibraries(false, progress);
-							} catch (final InvocationTargetException e) {
+								// externalLibraryWorkspace.updateState();
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/internal/N4JSExternalLibraryResourceDescriptionsPersisterContribution.java]: not needed?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/ExternalIndexSynchronizer.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	 * @return a map that maps projectIDs to their locations and versions
+	 */
+	final public Map<String, Pair<URI, String>> findNpmsInFolder() {
+		Map<String, Pair<URI, String>> npmsFolder = new HashMap<>();
+
+		java.net.URI nodeModulesLocation = locationProvider.getTargetPlatformNodeModulesLocation();
+		File nodeModulesFolder = new File(nodeModulesLocation.getPath());
+		if (nodeModulesFolder.exists() && nodeModulesFolder.isDirectory()) {
+			for (File npmLibrary : nodeModulesFolder.listFiles()) {
+				if (npmLibrary.exists() && npmLibrary.isDirectory()) {
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/ExternalIndexSynchronizer.java]: `.isDirectory()` is sufficient

**Comment by yoosiba** [comment]
this PR is resolving https://github.com/eclipse/n4js/issues/377 or not?

File: plugins/org.eclipse.n4js.n4mf/src/org/eclipse/n4js/n4mf/resource/N4MFResourceDescriptionStrategy.xtend [hunk 4]
```diff
@@ -117,6 +120,15 @@ class N4MFResourceDescriptionStrategy extends DefaultResourceDescriptionStrategy
 		builder.put(PROJECT_ID_KEY, projectId.nullToEmpty);
 		builder.put(IMPLEMENTATION_ID_KEY, implementationId.nullToEmpty);
 
+		val vers = projectVersion;
+		if (vers !== null) {
+			val versionStr = '''«vers.major».«vers.minor».«vers.micro»''';
```
[the remark this answers is not in the document]
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.n4mf/src/org/eclipse/n4js/n4mf/resource/N4MFResourceDescriptionStrategy.xtend]: unfortunately, I need this format here (`DeclaredVersionImpl.toString()` uses a different format)

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/EclipseExternalIndexSynchronizer.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+			case Install:
+			case Uninstall:
+				// nothing to do
+			}
+		}
+		return toBeDeleted;
+	}
+
+	private Set<URI> getToBeBuildProjects(Collection<LibraryChange> changeSet) {
```
[the remark this answers is not in the document]
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/EclipseExternalIndexSynchronizer.java]: thanks, but yeah...

**Pull request merged**